src/lib/core/prbuf.c

```diff
@@ -369,3 +376,170 @@ prbuf_iterator_next(struct prbuf_iterator *iter, struct prbuf_entry *result)
+				record_size_overhead >= header->size ||
+	    prbuf_header_size + header->end > header->size) {
+		diag_set(IllegalParams, "inconsistent header");
+		return -1;
+	}
+
+	return 0;
+}
+
+/** Read ahead is 128k. */
+#define PRBUF_READ_AHEAD (1 << 17)
+
+int
+prbuf_reader_create(struct prbuf_reader_iterator *iter, int fd, off_t offset)
+{
+	struct prbuf_header header;
+	/*
+	 * Read ahead does not make sense when reading prbuf header as
+	 * regularly first record is not at the beginning of the buffer.
+	 */
+	ssize_t rc = fio_pread(fd, &header, prbuf_header_size, offset);
+	if (rc == -1) {
+		diag_set(SystemError, "read failed");
+		return -1;
+	}
+	if ((size_t)rc < prbuf_header_size) {
+		diag_set(IllegalParams, "can't read buffer header");
+		return -1;
+	}
+	if (prbuf_check_header(&header) == -1)
+		return -1;
+
+	ibuf_create(&iter->buf, &cord()->slabc, PRBUF_READ_AHEAD);
+	iter->data_begin = offset + prbuf_header_size;
+	iter->data_end = offset + header.size;
+	iter->pos = iter->data_begin + header.begin;
+	iter->read_pos = iter->pos;
+	iter->fd = fd;
+
+	if (header.end == 0)
+		iter->unread_size = 0;
+	else if (header.begin < header.end)
+		iter->unread_size = header.end - header.begin;
+	else
+		iter->unread_size = header.size - prbuf_header_size -
+				    (header.begin - header.end);
+	return 0;
+}
+
+/**
+ * Read `size` data from file into `iter->buf`. Data is read ahead.
+ * `iter->read_pos` is increased accordingly.
+ *
+ * Besides IO error it is failure if we can't read requested amount of data.
+ *
+ * Return:
+ *   0 - success
+ *  -1 - failure (diag is set)
+ */
+static int
+prbuf_reader_ensure(struct prbuf_reader_iterator *iter, size_t size)
+{
+	if (ibuf_used(&iter->buf) >= size)
+		return 0;
+
+	size_t read_sz = size - ibuf_used(&iter->buf);
+	read_sz += PRBUF_READ_AHEAD;
+
+	char *buf = xibuf_reserve(&iter->buf, read_sz);
+	ssize_t rc = fio_pread(iter->fd, buf, read_sz, iter->read_pos);
+	if (rc < 0) {
+		diag_set(SystemError, "read failed");
+		return -1;
+	}
+	if ((size_t)rc < size) {
+		diag_set(IllegalParams, "can't read enough data");
+		return -1;
+	}
+	ibuf_alloc(&iter->buf, rc);
+	iter->read_pos += rc;
+	return 0;
+}
+
+/*
+ * Reset reading to the beginning of the prbuf data area.
+ */
+static inline void
+prbuf_reader_wrap(struct prbuf_reader_iterator *iter)
+{
+	iter->unread_size -= iter->data_end - iter->pos;
+	ibuf_reset(&iter->buf);
+	iter->pos = iter->data_begin;
+	iter->read_pos = iter->pos;
+}
+
+int
+prbuf_reader_next(struct prbuf_reader_iterator *iter,
+		  struct prbuf_entry *entry)
+{
+	if (iter->unread_size == 0) {
+		entry->ptr = NULL;
+		entry->size = 0;
+		return 0;
+	}
+
+	/* Check if we hit end of buffer and need to wrap around. */
+	assert(iter->data_end >= iter->pos);
+	if ((uint32_t)(iter->data_end - iter->pos) < record_size_overhead)
+		prbuf_reader_wrap(iter);
+
+	/* Read record length. */
+	int rc = prbuf_reader_ensure(iter, record_size_overhead);
+	if (rc == -1)
+		return -1;
```

Comment:
Nit: `rc` is unused. You could write:

```C
if (prbuf_reader_ensure(reader, size) != 0)
    return -1;
```

Same thing below.

Reply:
Done.